Update Jaeger CR

jpkrohling opened the pull request:
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jaegertracing/jaeger-operator/193)
<!-- Reviewable:end -->

jpkrohling commented:
This PR is an early draft, meant only to get feedback about the direction. 

The main changes here are:

- `strategy.S` is now a type, instead of an interface, and holds arrays of objects that compose that strategy (like, list of deployments, list of config maps)
- `newAllInOneStrategy` and similar methods now return a `strategy.S`
- the `handleCreate` and `handleUpdate` functions in the controller were removed and we now have an `apply` method, accepting the strategy. This calls more specialized functions, like `applyDeployments`, passing what's the desired state for `deployments`
- `applyDeployments` receives the desired state and compares with the current state by querying the cluster. It then builds an `Inventory`, which is composed of three arrays: `Create` (things that don't exist yet), `Update` (things that already exist) and `Delete` (things that no longer exist). 

Still missing:
- [ ] Polishing of the code
- [ ] Fix some broken tests and compilation issues
- [ ] Manual tests (what we have to far is only a draft of how it would work in theory)
- [ ] e2e tests

Things to decide:
- [ ] The code is more aware of the Kubernetes objects (Deployment vs. runtime.Object), which follows what we have in other parts of the code. I think we can make the code generic in a few places without major problems, but we'll end up having some duplication in some places. In the end, I think this is a fair price to pay, but wanted your opinions on that as well.

jpkrohling commented:
Good point. This validates then that going for concrete types (`Deployment`) has clear advantages against the generic `runtime.Object`.

jpkrohling commented:
I did some manual tests with this PR in its current state and I believe it's already at a state where we can consider merging, even if updating other components are missing (services, config maps, ...)

Here's what I did for testing:

```
$ kubectl apply -f deploy/examples/simplest.yaml 
jaeger.io.jaegertracing/simplest created


$ kubectl get pods
NAME                        READY     STATUS    RESTARTS   AGE
cassandra-0                 1/1       Running   0          5h
cassandra-1                 1/1       Running   3          5h
cassandra-2                 1/1       Running   0          5h
elasticsearch-0             1/1       Running   0          5h
simplest-6586f6bb95-mbsvv   1/1       Running   0          17m


$ kubectl apply -f deploy/examples/simplest.yaml  ### changed the Spec to be like the simple-prod
jaeger.io.jaegertracing/simplest configured


$ kubectl get pods
NAME                                  READY     STATUS        RESTARTS   AGE
cassandra-0                           1/1       Running       0          5h
cassandra-1                           1/1       Running       3          5h
cassandra-2                           1/1       Running       0          5h
elasticsearch-0                       1/1       Running       0          5h
simplest-6586f6bb95-mbsvv             0/1       Terminating   0          17m
simplest-collector-784544999c-bqnkp   0/1       Running       0          7s
simplest-query-77499c65d-2cp5c        1/1       Running       0          7s


$ kubectl get pods
NAME                                  READY     STATUS    RESTARTS   AGE
cassandra-0                           1/1       Running   0          5h
cassandra-1                           1/1       Running   3          5h
cassandra-2                           1/1       Running   0          5h
elasticsearch-0                       1/1       Running   0          5h
simplest-collector-784544999c-bqnkp   1/1       Running   0          10s
simplest-query-77499c65d-2cp5c        1/1       Running   0          10s


$ kubectl apply -f deploy/examples/simplest.yaml  #### back to the original simplest.yaml 
jaeger.io.jaegertracing/simplest configured


$ kubectl get pods
NAME                                  READY     STATUS              RESTARTS   AGE
cassandra-0                           1/1       Running             0          5h
cassandra-1                           1/1       Running             3          5h
cassandra-2                           1/1       Running             0          5h
elasticsearch-0                       1/1       Running             0          5h
simplest-6586f6bb95-bvs2r             0/1       ContainerCreating   0          2s
simplest-collector-784544999c-bqnkp   0/1       Terminating         0          53s
simplest-query-77499c65d-2cp5c        0/1       Terminating         0          53s


$ kubectl get pods
NAME                        READY     STATUS    RESTARTS   AGE
cassandra-0                 1/1       Running   0          5h
cassandra-1                 1/1       Running   3          5h
cassandra-2                 1/1       Running   0          5h
elasticsearch-0             1/1       Running   0          5h
simplest-6586f6bb95-bvs2r   0/1       Running   0          4s
```

objectiser commented:
@jpkrohling Looks promising, although main concern is that the new deployment is not confirmed as available before the old is removed - so there will be a time period where no traces are collected.

If we could solve this problem, then atleast if the user is using the same backend storage, it should be able to cope with continously collecting spans through the transition.

It also means that if the new deployment was to fail, we can abort and still have the old deployment available and continuing to collect.

One option is to deal with this scenario as a separate PR, which would be fine - as long as it was addressed soon.

jpkrohling commented:
>  Looks promising, although main concern is that the new deployment is not confirmed as available before the old is removed - so there will be a time period where no traces are collected.

Right, I think we discussed that but I missed it. I'll implement this next, still in this PR.

jpkrohling commented:
Marking as WIP, as there seems to be a problem with the unit test.

jpkrohling commented:
@objectiser this is now ready to be reviewed.

objectiser commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -97,26 +99,35 @@ func (r *ReconcileJaeger) Reconcile(request reconcile.Request) (reconcile.Result
+		log.WithFields(log.Fields{
+			"namespace": instance.Namespace,
+			"instance":  instance.Name,
+		}).WithError(err).Error("failed to update")
```
Does this message need to be more specific  - i.e. failed to store updated CR ? - as otherwise implies failed to update the jaeger instance (i.e. deployment)

objectiser commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -135,57 +146,32 @@ func defaultStrategyChooser(instance *v1alpha1.Jaeger) strategy.S {
-		// we probably want to add a couple of seconds to this deadline, but for now, this should be sufficient
-		deadline := time.Duration(*dep.Spec.ActiveDeadlineSeconds)
-		return wait.Poll(time.Second, deadline*time.Second, func() (done bool, err error) {
+		// default to 2 minutes, in case we get a null pointer
```
can you elaborate on the "in case we get a null pointer" comment?

jpkrohling replied:
`dep.Spec.ActiveDeadlineSeconds` is null if it's not specified. So far, only the Cassandra Create Schema job is a "dependency" and we control its deployment, so, it's always set, but I created a test without specifying this value and I got a null pointer panic.

objectiser commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -195,3 +181,94 @@ func (r *ReconcileJaeger) handleDependencies(str strategy.S) error {
+		}
+	}
+	for _, d := range depInventory.Update {
+		if err := r.waitForStability(d); err != nil {
```
Just wondering what action should be taken if a problem occurs? Not deleting the existing deployment is definitely a step in the right direction, but do we need to revert any changes that have been done (or remove any new deployments)? May be this has to be handled in a separate PR as more advanced handling?

jpkrohling replied:
Yeah, I think we need to think more about it during the advanced handling because it all depends on "when" it failed, as well as "where". For instance, if we successfully updated one deployment already, where do we get the previous definition for it, to roll it back?

objectiser commented on pkg/inventory/deployment_test.go:
```diff
@@ -0,0 +1,40 @@
+
+	depInventory := ForDeployments(existing, desired)
+	assert.Equal(t, "dep-to-create", depInventory.Create[0].Name)
+	assert.Len(t, depInventory.Create, 1)
```
nit: should test length before testing an element - just in case the list is empty. Same below.

objectiser commented on pkg/strategy/all-in-one_test.go:
```diff
@@ -61,24 +54,18 @@ func TestCreateAllInOneDeploymentWithUIConfigMap(t *testing.T) {
+	assertDeploymentsAndServicesForAllInOne(t, name, c, false, false, true)
 }
 
 func TestDelegateAllInOneDepedencies(t *testing.T) {
```
minor: TestDelegateAllInOneDependencies - also need update name in test.

objectiser commented on pkg/strategy/controller_test.go:
```diff
@@ -20,34 +17,24 @@ func TestNewControllerForAllInOneAsDefault(t *testing.T) {
 	ctrl := For(context.TODO(), jaeger)
-	ds := ctrl.Create()
-	assert.Len(t, ds, 6)
+	assert.Len(t, ctrl.Deployments(), 2)
```
Just wanted to check why the number of deployments has changed from 6 to 2?

jpkrohling replied:
Before, `Create()`  returned all the objects that should be created, like deployments, daemonsets, configmaps, and so on. `Deployments()` return only the deployments. `ds` was probably a bad name for that var.

jpkrohling replied:
By the way: I have not changed the test to check all the objects, as it's done already within the test for the concrete strategy. Instead, I just needed a quick way to ensure the code returns the right strategy. As we now have a `Type()` method, I changed the test to assert the strategy type instead.

objectiser commented on pkg/strategy/production_test.go:
```diff
@@ -118,11 +106,12 @@ func TestOptionsArePassed(t *testing.T) {
 
 func TestDelegateProductionDepedencies(t *testing.T) {
```
minor: spelling TestDelegateProductionDependencies and in the test

objectiser commented on pkg/strategy/streaming_test.go:
```diff
@@ -131,11 +119,12 @@ func TestStreamingOptionsArePassed(t *testing.T) {
 
 func TestDelegateStreamingDepedencies(t *testing.T) {
```
minor: same as in other tests - TestDelegateStreamingDependencies spelling and in test

objectiser commented:
@jpkrohling Tested by deploying `simplest.yaml` with the name field changed to `simple-prod`. Once running, I then applied `simple-prod.yaml` - it started the separate query and collector, but didn't terminate the all-in-one deployment.

jpkrohling commented:
As talked on IRC, this PR is missing the basic functionality of creating/updating non-deployment objects :-) I'm updating it with the other parts, but wanted to leave it for a short review while I fix a bug with config maps.

The second commit from this PR contains the changes to support other non-deployment objects (missing a few ones, WIP). 

We talked about having a generic vs. concrete approach and decided in favor of the concrete approach. The second PR shows that there's almost a 1-to-1 code duplication for the other resources, but I think the code is simple enough to justify this duplication. Making this code more generic would make things less flexible and would increase complexity, as there's no common interface between all the objects other than `runtime.Object`, which would require type checking whenever we need to use the object's name (as `runtime.Object` itself has no `Name`/`Namespace` properties).

Even though I believe the way it's done in this PR is the right choice, I'd like to get your opinion before moving forward.

jpkrohling commented:
The third commit in this PR shows what's the required code for individual object types (ingresses, in the example).

objectiser commented on pkg/controller/jaeger/ingress.go:
```diff
@@ -0,0 +1,59 @@
+		}
+	}
+
+	for _, d := range inv.Delete {
```
Just wondering if ingresses need to be handled similar to deployments - i.e. wait for new and updated to be stable before deleting old?

jpkrohling replied:
Just gave it a try, and ingresses are most likely not subject to many changes at all, as they point to services, like `simple-prod-query`:

```yaml
  spec:
    backend:
      serviceName: simple-prod-query
      servicePort: 16686
```

IIRC, we have only one ingress, and it's installed in all cases (except when running in OpenShift), so, I don't think we currently have the case where we have a new/updated ingress and an old one to be removed.

jpkrohling commented:
@pavolloffay : could you please review this one as @objectiser is away?

pavolloffay commented:
sure I will have a look. I would like to first fix the e2e tests so ES can be merged.

jpkrohling commented:
Rebasing to get the latest changes from ES broke this feature. I'm working on it, but would be nice to get a review started.

pavolloffay commented on pkg/account/oauth-proxy.go:
```diff
@@ -20,6 +20,14 @@ func OAuthProxy(jaeger *v1alpha1.Jaeger) *v1.ServiceAccount {
 			Namespace: jaeger.Namespace,
+			Labels: map[string]string{
+				"app":                          "jaeger",
+				"app.kubernetes.io/name":       OAuthProxyAccountNameFor(jaeger),
```
It would be great to have a util function which creates these labels, it seems like a repetitive patern

pavolloffay commented on pkg/controller/jaeger/account.go:
```diff
@@ -0,0 +1,59 @@
+
+	inv := inventory.ForAccounts(list.Items, desired)
+	for _, d := range inv.Create {
+		log.WithFields(log.Fields{
```
This again seems a repetitive and makes 30% of the file. We could use this https://github.com/sirupsen/logrus#default-fields and pass logger around. Or at least have a util function that returns logger with default fields.

jpkrohling replied:
This is how it would look like:

```golang
	logFields := log.WithFields(log.Fields{
		"namespace": jaeger.Namespace,
		"instance":  jaeger.Name,
	})
```

And then, within each loop iteration:

```golang
		logFields.WithField("account", d.Name).Debug("creating service account")
```

It does look better, just not sure how much better :) If you prefer it this way, I have no problems in changing it.

pavolloffay commented on pkg/controller/jaeger/configmap.go:
```diff
@@ -0,0 +1,59 @@
+
+	inv := inventory.ForConfigMaps(list.Items, desired)
+	for _, d := range inv.Create {
+		log.WithFields(log.Fields{
```
the same story about the logger

pavolloffay commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -97,26 +93,35 @@ func (r *ReconcileJaeger) Reconcile(request reconcile.Request) (reconcile.Result
 
 	// wait for all the dependencies to succeed
 	if err := r.handleDependencies(str); err != nil {
+		log.WithFields(log.Fields{
```
The same logging pattern

pavolloffay commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -135,63 +140,52 @@ func defaultStrategyChooser(instance *v1alpha1.Jaeger) strategy.S {
 	}
 
-	return nil
+	if viper.GetString("platform") == v1alpha1.FlagPlatformOpenShift {
```
strings.equalsfold?

pavolloffay commented:
@jpkrohling CI is failing. Could you please fix it? 

Do you have some test use cases which I should try?

jpkrohling commented:
> Do you have some test use cases which I should try?

e2e is coming, probably in a follow-up PR. For now, I'm testing by deploying the `simplest.yaml` changing its name to `simple-prod` and then deploying the `simple-prod.yaml`.

jpkrohling commented on deploy/examples/simplest.yaml:
```diff
@@ -1,4 +1,4 @@
 kind: Jaeger
 metadata:
-  name: simplest
+  name: simple-prod
```
Oops. Reverted.

jpkrohling commented on pkg/config/sampling/sampling.go:
```diff
@@ -55,6 +55,14 @@ func (u *Config) Get() *v1.ConfigMap {
 			Name:      fmt.Sprintf("%s-sampling-configuration", u.jaeger.Name),
 			Namespace: u.jaeger.Namespace,
+			Labels: map[string]string{
+				"app":                          "jaeger",
```
Something like this?

```golang
func CommonLabels(name string, instance string, component string) map[string]string {
	return map[string]string{
		"app":                          "jaeger",
		"app.kubernetes.io/name":       name,
		"app.kubernetes.io/instance":   instance,
		"app.kubernetes.io/component":  component,
		"app.kubernetes.io/part-of":    "jaeger",
		"app.kubernetes.io/managed-by": "jaeger-operator",
	}
}
```

pavolloffay replied:
+1 we should do the same for  owner reference - in a different PR

jpkrohling replied:
I'll create an issue and do both in a separate PR.

pavolloffay commented:
In the jaeger package there are very similar files for each type e.g. - deployment, daemonset, role. Especially the create update and delete sections - I think this should be dried out to a generic impl operating on runtime.object

jpkrohling commented:
> I think this should be dried out to a generic impl operating on runtime.object

I discussed it with @objectiser before and we consciously decided *not* to go this path.

pavolloffay commented on pkg/controller/jaeger/deployment.go:
```diff
@@ -0,0 +1,95 @@
+	// 3) deployments that are only on `existing` (delete)
+	depInventory := inventory.ForDeployments(depList.Items, desired)
+	for _, d := range depInventory.Create {
+		logFields.WithField("deployment", d.Name).Debug("creating deployment")
```
you pass the "deployment" twice into the log message. Maybe better to add it to fields.  I think this is in all files in this package.

jpkrohling replied:
You mean the word "deployment"?

pavolloffay replied:
yes

pavolloffay commented on pkg/controller/jaeger/role.go:
```diff
@@ -0,0 +1,52 @@
+
+	inv := inventory.ForRoles(list.Items, desired)
+	for _, d := range inv.Create {
+		logFields.WithField("account", d.Name).Debug("creating role")
```
The same comment as before:

* remane account to role and move it do fields. 
* remove role from debug.

jpkrohling replied:
Good catch on the naming problem, I just fixed that.

The role is part of the loop, it can't be moved to fields.

pavolloffay commented on pkg/controller/jaeger/jaeger_controller.go:
```diff
@@ -97,26 +96,23 @@ func (r *ReconcileJaeger) Reconcile(request reconcile.Request) (reconcile.Result
+	applied, err := r.apply(*instance, str)
 	if err != nil {
-		log.WithField("instance", instance).WithError(err).Error("failed to create")
+		logFields.WithError(err).Error("failed to apply the changes")
```
The error is returned below, Is it also being logged upstream? You might also wrap the error before returning.

jpkrohling replied:
This is the "entry point" for our application. It might be logged before, but this is the closed to the user as we have inside the operator.

pavolloffay commented on pkg/strategy/strategy.go:
```diff
@@ -0,0 +1,185 @@
+
+// S knows what type of deployments to build based on a given spec
+type S struct {
+	typ          Type
```
The type is used only in tests. Which could be check based on specific object -eg. all-in-one is present or injester

pavolloffay replied:
??

jpkrohling replied:
Sorry, missed this one. This is used in the `jaeger_controller`, inside the `ReconcileJaeger` struct

pavolloffay commented:
I have tried to deploy simplest and got:
```
Warning  FailedMount            9s (x8 over 1m)  kubelet, localhost  MountVolume.SetUp failed for volume "simplest-ui-oauth-proxy-tls" : secrets "simplest-ui-oauth-proxy-tls" not found
```

```
ploffay  ~/projects/golang/src/github.com/jaegertracing/jaeger-operator   PR193  make run  WATCH_NAMESPACE=myproject                                                                                                                              1:57 
customresourcedefinition.apiextensions.k8s.io/jaegers.io.jaegertracing created
INFO[0000] Versions                                      arch=amd64 operator-sdk=v0.4.1 os=linux version=go1.11.1
INFO[0000] Auto-detected the platform                    platform=openshift
INFO[0000] Starting the Cmd.                            
INFO[0334] Storage type wasn't provided for the Jaeger instance 'simplest'. Falling back to 'memory' 
ERRO[0335] failed to apply the changes                   error="Deployment.apps \"simplest\" not found" instance=simplest namespace=myproject

```

pavolloffay commented:
I have tried manually changing deployment e.g. `oc edit deployment ...`. In one case it failed on service account error showed in my previous comment. In the second try it created rs, and pods but then the pods were deleted, however the rs was still there.

I have also tried manually changing the service - e.g. port. Operator didn't revert this change.

jpkrohling commented:
I'll take a look at the failure in OpenShift. Looks like an annotation is missing somewhere. Thanks for the report!

About changing the deployments directly: this is not supported. The operator will override the changes whenever the reconciliation operation runs again (which might happen only when the CR is changed).

pavolloffay commented:
> About changing the deployments directly: this is not supported. The operator will override the changes whenever the reconciliation operation runs again (which might happen only when the CR is changed).

Could you please document this somewhere? Perhaps in the readme what updates are possible and what happens if somebody modifies objects directly.

pavolloffay commented:
>I'll take a look at the failure in OpenShift. Looks like an annotation is missing somewhere. Thanks for the report!

You are welcome! It seems that it happens only on OC 3.9.0
```
oc v3.9.0
kubernetes v1.9.1+cbc5b49
```

I haven't seen it on newer versions

pavolloffay commented:
However on new openshift versions I see this:
```
minishift v1.31.0+cfc599c


ERRO[0125] failed to apply the changes                   error="Operation cannot be fulfilled on services \"simplest-query\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Configured Jaeger instance                    instance=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0126] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0126] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simplest-ui-proxy\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[0127] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0127] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0127] Reconciling Deployment                        name=simplest namespace=myproject
INFO[0127] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0127] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simplest\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[0128] Reconciling Deployment    
```

The log is repeating and repeating.

pavolloffay commented:
I get following error when deploying `oc create -f deploy/examples/simple-prod-deploy-es.yaml` on minishift
```
INFO[6101] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[6101] failed to store back the current CustomResource  error="Operation cannot be fulfilled on jaegers.io.jaegertracing \"simplest\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[6102] Reconciling Deployment                        name=simplest namespace=myproject
ERRO[6144] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6145] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6146] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6148] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6149] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6150] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6151] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6152] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simple-prod-ui-proxy\": the object has been modified; please apply your changes to the latest version and try again" instance=simple-prod namespace=myproject
ERRO[6154] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6155] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6158] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6164] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERRO[6174] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproject
ERROR: logging before flag.Parse: W0222 16:31:39.069050   22201 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:196: watch of *v1.ConfigMap ended with: too old resource version: 223069 (223162)
ERRO[6195] failed to apply the changes                   error="CronJob.batch \"simple-prod-es-index-cleaner\" is invalid: spec.jobTemplate.spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"certs\"" instance=simple-prod namespace=myproj
```

jpkrohling commented:
I fixed the OpenShift problems (v3.11), but there's still one point open that might be a bug in the operator SDK. When running in Kubernetes, the reconcile loop is being called only when the CR is changed, while in OpenShift it's being called every second. Not sure yet why, but I opened a thread in the operator-framework mailing list addressing that.

I also addressed the comments from the last review, except one (left a question on the item itself).

pavolloffay commented:
It would be good to add a link to operator ML to this thread. 

Do you want me to re-review?

jpkrohling commented:
> It would be good to add a link to operator ML to this thread.

https://groups.google.com/d/msg/operator-framework/DZynMacOgoY/RWaSFxFUBwAJ

I'll track this in a separate issue so that it won't block this PR (#203 is probably related)

> Do you want me to re-review?

Yes, please.

pavolloffay commented on pkg/controller/jaeger/account.go:
```diff
@@ -0,0 +1,52 @@
+
+	inv := inventory.ForAccounts(list.Items, desired)
+	for _, d := range inv.Create {
+		logFields.WithField("account", d.Name).Debug("creating service account")
```
move `logFields.WithField("account", d.Name)` to fields above and `s/creating service account/creating/`

pavolloffay replied:
Do the same in all new  files in this PR

jpkrohling replied:
I'll do the documentation, but the logging is a matter of taste... I'll keep it like it is.

pavolloffay replied:
:+1: good to know

pavolloffay commented on pkg/controller/jaeger/deployment_test.go:
```diff
@@ -0,0 +1,197 @@
+	}()
+
+	// we assume that this sleep time is enough for the reconcile to reach the "wait" logic
+	time.Sleep(100 * time.Millisecond)
```
Could be this we avoided completelly?

jpkrohling replied:
Unfortunately, no. I didn't have it before and it caused the tests to intermittently fail, as it's not certain that the "wait" part inside the reconcile loop would have started already.

Basically, what we want is the reconcile loop to run and create the deployment/cron job and wait until it reaches the stability / completes the job and we want the "waitFor" to execute at least once. *Then* we are ready to update the "available"/"succeeded" fields and the wait should finish, reaching the `wg.Done()` part.

pavolloffay commented on pkg/controller/jaeger/secret_test.go:
```diff
@@ -0,0 +1,131 @@
+	assert.NoError(t, err)
+}
+
+func TestSecretsDelete(t *testing.T) {
```
The tests in this package look very similar. Possibly to dry them out?

jpkrohling replied:
Same problem as with the main code: either we add complexity by making the code generic, or we accept the duplication and keep the code simple/flexible.

pavolloffay commented:
running on minishift produces following errors. THe serviceaccount error repeates over time.
```
 ✘ 130 ploffay  ~/projects/golang/src/github.com/jaegertracing/jaeger-operator   PR193  make run WATCH_NAMESPACE=myproject                                                                                                                        11:33 
INFO[0000] Versions                                      arch=amd64 operator-sdk=v0.4.1 os=linux version=go1.11.1
INFO[0000] Auto-detected the platform                    platform=openshift
INFO[0000] Starting the Cmd.                            
INFO[0010] Storage type wasn't provided for the Jaeger instance 'simplest'. Falling back to 'memory' 
ERRO[0011] failed to apply the changes                   error="Deployment.apps \"simplest\" not found" instance=simplest namespace=myproject
INFO[0012] Storage type wasn't provided for the Jaeger instance 'simplest'. Falling back to 'memory' 
INFO[0024] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0024] failed to apply the changes                   error="routes.route.openshift.io \"simplest\" already exists" instance=simplest namespace=myproject
INFO[0025] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0025] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simplest\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[0026] Configured Jaeger instance                    instance=simplest namespace=myproject
INFO[0026] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0026] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simplest-ui-proxy\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
INFO[0027] Configured Jaeger instance                    instance=simplest namespace=myproject
ERRO[0027] failed to apply the changes                   error="Operation cannot be fulfilled on serviceaccounts \"simplest-ui-proxy\": the object has been modified; please apply your changes to the latest version and try again" instance=simplest namespace=myproject
```

jpkrohling commented:
> running on minishift produces following errors. THe serviceaccount error repeates over time.

Are you running on a clean OpenShift, or could this be something left behind from previous iterations? I do get the service account one from time to time, but should be relatively innocuous, once #231 is fixed.

pavolloffay commented:
It's clean OpenShift. The #231 causes the error to pop up more ofter, maybe there is something wrong with SA bc other objects are not logged with the same modified error.

pavolloffay commented:
The problem is that `ResourceVersion` is SA is always different as it runs the loop. Adding `tp.ResourceVersion = v.ResourceVersion` to inventory/sa fixes the problem - although I am not sure whether it's correct to do so.

jpkrohling commented:
The resource version field is the reason we update individual fields :) We are not supposed to touch that one: https://groups.google.com/forum/#!topic/operator-framework/KqgrZ3v2P9U

jpkrohling commented:
> They do tests change in annotations - basically `ObjectMeta` which fine but they should more importantly tests if core fields of objects get updated.

I added a few more tests for fields within the `Spec` (or related) objects.

pavolloffay commented:
> The resource version field is the reason we update individual fields :) We are not supposed to touch that one: https://groups.google.com/forum/#!topic/operator-framework/KqgrZ3v2P9U

thanks for the pointer. There is definitelly something fishy with SA or secrets. I have noticed that every loop creates new secrets.

```
oc get secrets | wc -l                                                                                                                                           1:54 
597
```

jpkrohling commented:
> There is definitelly something fishy with SA or secrets. I have noticed that every loop creates new secrets.

That's a great hint! I did not notice this before, let me see if I can figure out what's wrong.

jpkrohling commented:
So, I tracked the problem down to the reconcile logic, where we store back the CR as we might have done some reconciliation. Once that is done, I don't see #231 anymore and the number of service accounts becomes stable.

I'll track that in a different issue, to unblock this PR.

jpkrohling commented:
e2e tests are passing, will merge this:

```
$ make test
Running unit tests...
?   	github.com/jaegertracing/jaeger-operator/cmd	[no test files]
?   	github.com/jaegertracing/jaeger-operator/cmd/manager	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/account	0.017s	coverage: 100.0% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/apis	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/apis/io/v1alpha1	0.037s	coverage: 15.8% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/cmd/start	[no test files]
?   	github.com/jaegertracing/jaeger-operator/pkg/cmd/version	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/config/sampling	0.007s	coverage: 94.1% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/config/ui	0.015s	coverage: 94.1% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/controller	[no test files]
?   	github.com/jaegertracing/jaeger-operator/pkg/controller/deployment	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/controller/jaeger	2.339s	coverage: 73.9% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/cronjob	0.080s	coverage: 93.9% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/deployment	0.079s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/ingress	0.101s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/inject	0.026s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/inventory	0.059s	coverage: 91.8% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/route	0.059s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/service	0.040s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/storage	0.631s	coverage: 93.2% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/storage/elasticsearch/v1alpha1	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/strategy	0.015s	coverage: 92.1% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/util	0.010s	coverage: 100.0% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/version	[no test files]
Formatting code...
Building...
Sending build context to Docker daemon  130.8MB
Step 1/6 : FROM centos
 ---> 1e1148e4cc2c
Step 2/6 : RUN INSTALL_PKGS="       openssl       " &&     yum install -y $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     yum clean all &&     mkdir /tmp/_working_dir &&     chmod og+w /tmp/_working_dir
 ---> Using cache
 ---> 82728e7bfc7b
Step 3/6 : COPY scripts/* /scripts/
 ---> Using cache
 ---> 42400106dc1c
Step 4/6 : USER nobody
 ---> Using cache
 ---> a590f9fcf745
Step 5/6 : ADD build/_output/bin/jaeger-operator /usr/local/bin/jaeger-operator
 ---> 55e2e02811d3
Step 6/6 : ENTRYPOINT ["/usr/local/bin/jaeger-operator"]
 ---> Running in 666da8612545
Removing intermediate container 666da8612545
 ---> 1fc614a5959e
Successfully built 1fc614a5959e
Successfully tagged jpkroehling/jaeger-operator:latest
Pushing image jpkroehling/jaeger-operator:latest...
Running end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	371.189s
```

pavolloffay commented on pkg/strategy/production.go:
```diff
@@ -22,98 +20,96 @@ import (
-			Jaeger: c.jaeger,
+			Jaeger: jaeger,
 		}
-		objs, err := es.CreateElasticsearchObjects(cDep.Spec.Template.Spec.ServiceAccountName, queryDep.Spec.Template.Spec.ServiceAccountName)
```
@jpkrohling this PR broke Elasticsearch functionality, note that `es.CreateElasticsearchObjects` was returning ES CR in the object array. Now the ES CR is not being processed.
